lib/cretonne/src/licm.rs

```diff
@@ -15,40 +20,137 @@ pub fn do_licm(func: &mut Function,
+    // loop, then we look at its arguments: if they are not defined inside the same loop, then the
+    // instruction is loop-invariant and we hoist it. We don't have to test for child loop because
+    // of the reverse postorder and the immediate hoisting guarantees that later instructions that
+    // use its result value can also be hoisted.
+    for &inst in insts_rpo.as_slice(&inst_pool) {
+        // First which loop is the instruction in
+        let inst_loop = match loop_analysis.base_loop_inst(inst, &func.layout) {
+            None => continue,
+            Some(lp) => lp,
+        };
+        // First we find where is the closest definition of one of the instruction's arguments
+        let closest_def_loop = closest_def_position(func, inst, inst_loop, loop_analysis);
+        // If all the instruction's argument are defined outside the loop then this instruction is
+        // loop-invariant
+        let destination_loop = match closest_def_loop {
+            // If all the arguments are defined outside any loops we move the instruction just
+            // before the outermost loop containing inst_loop.
+            DefPosition::OutsideLoop() => loop_analysis.outermost_loop(inst_loop, None),
+            // In this case we move just before the loop which would be a direct child of lp
+            DefPosition::ParentLoop(lp) => loop_analysis.outermost_loop(inst_loop, Some(lp)),
+            // If an argument is defined inside the current loop, we abort because the instruction
+            // is not loop-invariant.
+            DefPosition::InsideCurrentLoop() => continue,
+        };
+        match has_pre_header(&func.layout,
+                             cfg,
+                             domtree,
+                             loop_analysis.loop_header(destination_loop)) {
+            None => {
+                let loop_header = loop_analysis.loop_header(destination_loop);
+                let (pre_header, jump_inst) = create_pre_header(loop_header, func, cfg, domtree);
+                loop_analysis.recompute_loop_preheader(pre_header, loop_header);
+                domtree.recompute_loop_preheader(loop_header, pre_header, jump_inst);
+                {
+                    let mut cur = Cursor::new(&mut func.layout);
+                    cur.goto_inst(inst);
+                    cur.remove_inst();
+                    cur.goto_inst(jump_inst);
+                    cur.insert_inst(inst);
+                }
+            }
+            // If there is a natural pre-header we insert new instructions just before the
+            // related jumping instruction (which is not necessarily at the end).
+            Some((_, last_jump_inst)) => {
+                let mut cur = Cursor::new(&mut func.layout);
+                cur.goto_inst(inst);
+                cur.remove_inst();
+                cur.goto_inst(last_jump_inst);
+                cur.insert_inst(inst);
+            }
```

Comment:
This code is pretty verbose for doing a conceptually simple task, moving an inst, and it appears twice. I suggest factoring this out into a helper function, either on layout or just a local function.